# 1. Autocorrelation coefficients

Autocorrelation is the correlation between a variable lagged one or
more time periods and itself.

Q1.Use Excel to compute the autocorrelation coefficients and display the autocorrelation function for the data presented in Table 1.

In [2]:
import pandas as pd

df = pd.DataFrame (
    {
        "Month":["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"],
        "VCR":[123,130,125,138,145,142,141,146,147,157,150,160]
    }
)

print(df)

   Month  VCR
0    Jan  123
1    Feb  130
2    Mar  125
3    Apr  138
4    May  145
5    Jun  142
6    Jul  141
7    Aug  146
8    Sep  147
9    Oct  157
10   Nov  150
11   Dec  160


![title](img\img1.png)

In [3]:
df["Yt_1"]=df["VCR"].shift(1)
df["Yt_2"]=df["VCR"].shift(2)
df["Yt_3"]=df["VCR"].shift(3)
df

Month  VCR   Yt_1   Yt_2   Yt_3
0    Jan  123    NaN    NaN    NaN
1    Feb  130  123.0    NaN    NaN
2    Mar  125  130.0  123.0    NaN
3    Apr  138  125.0  130.0  123.0
4    May  145  138.0  125.0  130.0
5    Jun  142  145.0  138.0  125.0
6    Jul  141  142.0  145.0  138.0
7    Aug  146  141.0  142.0  145.0
8    Sep  147  146.0  141.0  142.0
9    Oct  157  147.0  146.0  141.0
10   Nov  150  157.0  147.0  146.0
11   Dec  160  150.0  157.0  147.0

In [4]:
# mean Yt_k.

Ybar = df["VCR"].mean()
sum_Yt = df["VCR"].sum()
mean_Yt_1 = df["Yt_1"].mean()
mean_Yt_2 = df["Yt_2"].mean()
mean_Yt_3 = df["Yt_3"].mean()

print(sum_Yt)
print(Ybar)
print(mean_Yt_1)
print(mean_Yt_2)
print(mean_Yt_3)

1704
142.0
140.36363636363637
139.4
137.44444444444446


In [5]:
df["Yt-Ybar"] = df["VCR"]-Ybar
df["Yt_1-Ybar"] = df["Yt_1"]-Ybar
df["Yt_2-Ybar"] = df["Yt_2"]-Ybar
df["Yt_3-Ybar"] = df["Yt_3"]-Ybar

df

Month  VCR   Yt_1   Yt_2   Yt_3  Yt-Ybar  Yt_1-Ybar  Yt_2-Ybar  Yt_3-Ybar
0    Jan  123    NaN    NaN    NaN    -19.0        NaN        NaN        NaN
1    Feb  130  123.0    NaN    NaN    -12.0      -19.0        NaN        NaN
2    Mar  125  130.0  123.0    NaN    -17.0      -12.0      -19.0        NaN
3    Apr  138  125.0  130.0  123.0     -4.0      -17.0      -12.0      -19.0
4    May  145  138.0  125.0  130.0      3.0       -4.0      -17.0      -12.0
5    Jun  142  145.0  138.0  125.0      0.0        3.0       -4.0      -17.0
6    Jul  141  142.0  145.0  138.0     -1.0        0.0        3.0       -4.0
7    Aug  146  141.0  142.0  145.0      4.0       -1.0        0.0        3.0
8    Sep  147  146.0  141.0  142.0      5.0        4.0       -1.0        0.0
9    Oct  157  147.0  146.0  141.0     15.0        5.0        4.0       -1.0
10   Nov  150  157.0  147.0  146.0      8.0       15.0        5.0        4.0
11   Dec  160  150.0  157.0  147.0     18.0        8.0       15.0        5.0

In [6]:
df["(Yt-Ybar)(Yt_1-Ybar)"]=df["Yt-Ybar"]*df["Yt_1-Ybar"]
df["(Yt-Ybar)(Yt_2-Ybar)"]=df["Yt-Ybar"]*df["Yt_2-Ybar"]
df["(Yt-Ybar)(Yt_3-Ybar)"]=df["Yt-Ybar"]*df["Yt_3-Ybar"]
df["(Yt-Ybar)^2"]=df["Yt-Ybar"]**2
df

Month  VCR   Yt_1   Yt_2   Yt_3  Yt-Ybar  Yt_1-Ybar  Yt_2-Ybar  Yt_3-Ybar  \
0    Jan  123    NaN    NaN    NaN    -19.0        NaN        NaN        NaN   
1    Feb  130  123.0    NaN    NaN    -12.0      -19.0        NaN        NaN   
2    Mar  125  130.0  123.0    NaN    -17.0      -12.0      -19.0        NaN   
3    Apr  138  125.0  130.0  123.0     -4.0      -17.0      -12.0      -19.0   
4    May  145  138.0  125.0  130.0      3.0       -4.0      -17.0      -12.0   
5    Jun  142  145.0  138.0  125.0      0.0        3.0       -4.0      -17.0   
6    Jul  141  142.0  145.0  138.0     -1.0        0.0        3.0       -4.0   
7    Aug  146  141.0  142.0  145.0      4.0       -1.0        0.0        3.0   
8    Sep  147  146.0  141.0  142.0      5.0        4.0       -1.0        0.0   
9    Oct  157  147.0  146.0  141.0     15.0        5.0        4.0       -1.0   
10   Nov  150  157.0  147.0  146.0      8.0       15.0        5.0        4.0   
11   Dec  160  150.0  157.0  147.0     18.0        8.0       15.0        5.0   

    (Yt-Ybar)(Yt_1-Ybar)  (Yt-Ybar)(Yt_2-Ybar)  (Yt-Ybar)(Yt_3-Ybar)  \
0                    NaN                   NaN                   NaN   
1                  228.0                   NaN                   NaN   
2                  204.0                 323.0                   NaN   
3                   68.0                  48.0                  76.0   
4                  -12.0                 -51.0                 -36.0   
5                    0.0                  -0.0                  -0.0   
6                   -0.0                  -3.0                   4.0   
7                   -4.0                   0.0                  12.0   
8                   20.0                  -5.0                   0.0   
9                   75.0                  60.0                 -15.0   
10                 120.0                  40.0                  32.0   
11                 144.0                 270.0                  90.0   

    (Yt-Ybar)^2  
0         361.0  
1         144.0  
2         289.0  
3          16.0  
4           9.0  
5           0.0  
6           1.0  
7          16.0  
8          25.0  
9         225.0  
10         64.0  
11        324.0

In [7]:
r1=df["(Yt-Ybar)(Yt_1-Ybar)"].sum()/df["(Yt-Ybar)^2"].sum()
r2=df["(Yt-Ybar)(Yt_2-Ybar)"].sum()/df["(Yt-Ybar)^2"].sum()
r3=df["(Yt-Ybar)(Yt_3-Ybar)"].sum()/df["(Yt-Ybar)^2"].sum()
print(r1)
print(r2)
print(r3)

0.5719131614654003
0.4626865671641791
0.11058344640434192


### Results 
Positive lag 1 autocorrelation exists in this time series. The correlation between Yt and Yt-1
or the autocorrelation for time lag 1, is .572. This means that the successive monthly sales of VCRs are somewhat correlated with each other.

It appears that moderate autocorrelation exists in this time series lagged two time periods. The correlation between Yt and Yt-2
, or the autocorrelation for time lag 2, is .463.

The autocorrelation for time lag 3, is .111. This means a weak autocorrelation exists in this
time series lagged three time periods.

Notice that the autocorrelation coefficient at time lag 2 (.463) is less than the
autocorrelation coefficient at time lag 1 (.572). Generally, as the number of time lags ( k )
increases, the magnitudes of the autocorrelation coefficients decrease.

---

# 2. Are the Data Random?

- If a series is random, the autocorrelations between Yt
and Yt-k
for any
time lag k are close to zero. The successive values of a time series are
not related to each other.


## 2.1 To determine whether an autocorrelation coefficient is significantly different from zero

- Compute the standard deviations (or standard errors) of the autocorrelation coefficients.

- Compute the Ljung-Box Q statistics to examine a set of consecutive rk’s as a group.

## 2.2 Do the Data have a Trend?
- If a series has a trend, successive observations are highly correlated,
and the auto- correlation coefficients typically are significantly
different from zero for the first several time lags and then gradually
drop toward zero as the number of lags increases. The
autocorrelation coefficient for time lag 1 will often be very large
(close to 1). The autocorrelation coefficient for time lag 2 will also be
large. However, it will not be as large as for time lag 1.

## 2.3 Are the Data Seasonal?
- If a series has a seasonal pattern, a significant autocorrelation
coefficient will occur at the seasonal time lag or multiples of the
seasonal lag. The seasonal lag is 4 for quarterly data and 12 for
monthly data.

---

# 3.Choosing a Forecasting Technique
![title](img\ForecastingTechnique.png)

## 3.1 Measuring Forecasting 
![title](img\methods.png)

![title](img\errorSample.png)

Problem 1

An analyst would like to determine whether there is a pattern to earnings per share for the Price Company, which operated a wholesale/retail cash-and-carry business in several states under the name Price Club. The data are shown in Table P-20. (You may download the data file from Moodle)

a) i) Describe any patterns that exist in these data.

ii) Find the forecast value of the quarterly earnings per share for Price Club

for each quarter by using the naive approach (i.e., the forecast for firstquarter 1987 is the value for fourth-quarter 1986, .32).

b) Evaluate the naive forecast using MAD.

c) Evaluate the naive forecast using MSE and RMSE.

d) Evaluate the naive forecast using MAPE.

e) Evaluate the naive forecast using MPE. 

In [20]:
df = pd.DataFrame(
    {
    "Yt":[0.4,0.29,0.24,0.32,0.47,0.34,0.3,0.39,0.63,0.43,0.38,0.49,0.76,0.51,0.42,0.61,0.86,0.51,0.47,0.63,0.94,0.56,0.5,0.65,0.95,0.42,0.57,0.6,0.93,0.38,0.37,0.57
]
    }
)
df

Yt
0   0.40
1   0.29
2   0.24
3   0.32
4   0.47
5   0.34
6   0.30
7   0.39
8   0.63
9   0.43
10  0.38
11  0.49
12  0.76
13  0.51
14  0.42
15  0.61
16  0.86
17  0.51
18  0.47
19  0.63
20  0.94
21  0.56
22  0.50
23  0.65
24  0.95
25  0.42
26  0.57
27  0.60
28  0.93
29  0.38
30  0.37
31  0.57

In [23]:
# focast value
df["YtCap"] = df["Yt"].shift(1)
# error
df["Yt-YtCap"] = df["Yt"]-df["YtCap"] 
#abs error
df["abs(Yt-YtCap)"] = df["Yt-YtCap"].abs()
#error ^ 2
df["(Yt-YtCap)^2"] = df["Yt-YtCap"]**2
#abs error /abs true value
df["abs(Yt-YtCap)/abs(Yt)"] = df["abs(Yt-YtCap)"]/df["Yt"].abs()
#error / true value
df["(Yt-YtCap)/Yt"] = df["Yt-YtCap"]/df["Yt"]
df

Yt  YtCap  Yt-YtCap  abs(Yt-YtCap)  (Yt-YtCap)^2  abs(Yt-YtCap)/abs(Yt)  \
0   0.40    NaN       NaN            NaN           NaN                    NaN   
1   0.29   0.40     -0.11           0.11        0.0121               0.379310   
2   0.24   0.29     -0.05           0.05        0.0025               0.208333   
3   0.32   0.24      0.08           0.08        0.0064               0.250000   
4   0.47   0.32      0.15           0.15        0.0225               0.319149   
5   0.34   0.47     -0.13           0.13        0.0169               0.382353   
6   0.30   0.34     -0.04           0.04        0.0016               0.133333   
7   0.39   0.30      0.09           0.09        0.0081               0.230769   
8   0.63   0.39      0.24           0.24        0.0576               0.380952   
9   0.43   0.63     -0.20           0.20        0.0400               0.465116   
10  0.38   0.43     -0.05           0.05        0.0025               0.131579   
11  0.49   0.38      0.11           0.11        0.0121               0.224490   
12  0.76   0.49      0.27           0.27        0.0729               0.355263   
13  0.51   0.76     -0.25           0.25        0.0625               0.490196   
14  0.42   0.51     -0.09           0.09        0.0081               0.214286   
15  0.61   0.42      0.19           0.19        0.0361               0.311475   
16  0.86   0.61      0.25           0.25        0.0625               0.290698   
17  0.51   0.86     -0.35           0.35        0.1225               0.686275   
18  0.47   0.51     -0.04           0.04        0.0016               0.085106   
19  0.63   0.47      0.16           0.16        0.0256               0.253968   
20  0.94   0.63      0.31           0.31        0.0961               0.329787   
21  0.56   0.94     -0.38           0.38        0.1444               0.678571   
22  0.50   0.56     -0.06           0.06        0.0036               0.120000   
23  0.65   0.50      0.15           0.15        0.0225               0.230769   
24  0.95   0.65      0.30           0.30        0.0900               0.315789   
25  0.42   0.95     -0.53           0.53        0.2809               1.261905   
26  0.57   0.42      0.15           0.15        0.0225               0.263158   
27  0.60   0.57      0.03           0.03        0.0009               0.050000   
28  0.93   0.60      0.33           0.33        0.1089               0.354839   
29  0.38   0.93     -0.55           0.55        0.3025               1.447368   
30  0.37   0.38     -0.01           0.01        0.0001               0.027027   
31  0.57   0.37      0.20           0.20        0.0400               0.350877   

    (Yt-YtCap)/Yt  
0             NaN  
1       -0.379310  
2       -0.208333  
3        0.250000  
4        0.319149  
5       -0.382353  
6       -0.133333  
7        0.230769  
8        0.380952  
9       -0.465116  
10      -0.131579  
11       0.224490  
12       0.355263  
13      -0.490196  
14      -0.214286  
15       0.311475  
16       0.290698  
17      -0.686275  
18      -0.085106  
19       0.253968  
20       0.329787  
21      -0.678571  
22      -0.120000  
23       0.230769  
24       0.315789  
25      -1.261905  
26       0.263158  
27       0.050000  
28       0.354839  
29      -1.447368  
30      -0.027027  
31       0.350877

In [1]:
#mean absolute deviation (abs average error  )
MAD = df["abs(Yt-YtCap)"].sum()/df["YtCap"].count()
#mean squared error
MSE = df["(Yt-YtCap)^2"].sum()/df["YtCap"].count()
#root mean squared error
RMSE = MSE ** (1/2)
#mean absolute percentage error
MAPE = df["abs(Yt-YtCap)/abs(Yt)"].sum()/df["YtCap"].count()
#mean percentage error
MPE = df["(Yt-YtCap)/Yt"].sum()/df["YtCap"].count()
print(MAD , MSE , RMSE , MAPE , MPE)

NameError: name 'df' is not defined

32